File: 2022/day-10.rb
def parse(l)
  if l.start_with?("addx")
    [2, l.sub("addx ", "").to_i]
  else
    [1]
  end
end

input = File.read('day-10-input.txt')
  .split("\n")
  .map {|l| parse l }




test = DATA.map {|l| parse l }

def preprocess(cmds)
  cmds.inject([[0, 1]]) do |r, cmd|
    r << [r.last.first + cmd.first, r.last[1] + cmd[1].to_i, *cmd]
    r
  end
end

def p(cmds)
  (20..220).step(40).map do |cycle|
    x = cmds.take_while {|cmd| cmd.first < cycle}.last
    x[1] * cycle
  end
end



result1 = preprocess input
puts "Part 1: #{ p(result1).sum }"


def at_cycle(cycle, cmds)
  x = cmds.take_while {|cmd| cmd.first < cycle}.last

  sprite_pos = x[1]
  r = (sprite_pos-1..sprite_pos+1).include?((cycle % 40) - 1) ? '@' : ' '

  # puts "cycle #{cycle.to_s.rjust(3, ' ')} #{r} pos: #{sprite_pos}  #{x}"
  r
end



cmds = preprocess input
pixels = (1..240).map {|cycle| at_cycle(cycle, cmds) }
# puts "Part 2: #{ pixels }"

pixels.each_slice(40) do |l|
  puts l.join
end


__END__
addx 15
addx -11
addx 6
addx -3
addx 5
addx -1
addx -8
addx 13
addx 4
noop
addx -1
addx 5
addx -1
addx 5
addx -1
addx 5
addx -1
addx 5
addx -1
addx -35
addx 1
addx 24
addx -19
addx 1
addx 16
addx -11
noop
noop
addx 21
addx -15
noop
noop
addx -3
addx 9
addx 1
addx -3
addx 8
addx 1
addx 5
noop
noop
noop
noop
noop
addx -36
noop
addx 1
addx 7
noop
noop
noop
addx 2
addx 6
noop
noop
noop
noop
noop
addx 1
noop
noop
addx 7
addx 1
noop
addx -13
addx 13
addx 7
noop
addx 1
addx -33
noop
noop
noop
addx 2
noop
noop
noop
addx 8
noop
addx -1
addx 2
addx 1
noop
addx 17
addx -9
addx 1
addx 1
addx -3
addx 11
noop
noop
addx 1
noop
addx 1
noop
noop
addx -13
addx -19
addx 1
addx 3
addx 26
addx -30
addx 12
addx -1
addx 3
addx 1
noop
noop
noop
addx -9
addx 18
addx 1
addx 2
noop
noop
addx 9
noop
noop
noop
addx -1
addx 2
addx -37
addx 1
addx 3
noop
addx 15
addx -21
addx 22
addx -6
addx 1
noop
addx 2
addx 1
noop
addx -10
noop
noop
addx 20
addx 1
addx 2
addx 2
addx -6
addx -11
noop
noop
noop

File: 2022/day-10.hs
import Flow
import Data.List (isPrefixOf, stripPrefix)
import Data.List.Split (splitOn)
import Data.Map (Map, fromList, (!))
import Debug.Trace (trace)

filterEmptyString :: [String] -> [String]
filterEmptyString = filter ((> 0) . length)

data Command = Noop | AddX Integer deriving (Show)
data ExecResult = ExecResult { clock :: Integer
                             , x :: Integer
                             } deriving (Show)

addXPrefix = "addx "

parseAddX :: String -> Command
parseAddX s = AddX (read s'') where
  s' = stripPrefix addXPrefix s
  s'' = case s' of
    Just v  -> v
    Nothing -> error "wut?"

parseCommand :: String -> Command
parseCommand s
  | isPrefixOf addXPrefix s = parseAddX s
  | True = Noop

parseInput :: String -> [[Char]]
parseInput = filterEmptyString . splitOn "\n"

exec :: Command -> ExecResult -> ExecResult
exec (Noop) (ExecResult {x=x, clock=clock}) = ExecResult {x=x, clock=clock+1}
exec (AddX dx) (ExecResult {x=x, clock=clock}) = ExecResult {x=x+dx, clock=clock+2}

atClock :: Integer -> [ExecResult] -> ExecResult
atClock c = head . reverse . takeWhile (\(ExecResult {clock=clock}) -> clock < c)

pixelAt :: Integer -> [ExecResult] -> Char
-- pixelAt c results = trace (show (c, output, c', m, n, "x: " ++ (show x), aligned)) $ output
pixelAt c results = output
            where
  ExecResult { x=x, clock=clock} = atClock c results
  c' = (mod c 40) - 1
  m = (mod (x-1) 40)
  m' = (mod x 40)
  n = (mod (x+1) 40) -- assuming that the sprit will never wrap to the next line
  -- aligned = m <= c' && c' <= n
  aligned = c' == m || c' == m' || c' == n
  output = if aligned
           then '#'
           else ' '

display [] = return ()
display l = do
  putStrLn (take 40 l)
  display (drop 40 l)


main = do
  contents <- readFile "day-10-input.txt"
  let input = contents |> parseInput |> map parseCommand

  -- scan input ::
  let dat = foldl (\(lastResult:results) cmd -> (exec cmd lastResult):lastResult:results) [(ExecResult {x=1, clock=0})] input |> reverse

  print $ "Part 1: " ++ show input
  print $ "Part 1: " ++ show dat


  let y = map (\c -> (x (atClock (c) dat)) * (c)) [20,60..220]

  print $ "Part 1: " ++ show (sum y)



  let pixels = map (flip pixelAt dat) [1..240]

  print $ "Part 2: "

  -- display pixels
  display pixels
